src/libraries/Microsoft.Extensions.Options/src/ValidateOptionsResultBuilder.cs

```diff
@@ -0,0 +1,127 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Extensions.Options
+{
+    /// <summary>
+    /// Builds <see cref="ValidateOptionsResult"/> with support for multiple error messages.
+    /// </summary>
+    [DebuggerDisplay("{ErrorsCount} errors")]
+    public class ValidateOptionsResultBuilder
+    {
+        private const string MemberSeparatorString = ", ";
+
+        private List<string>? _errors;
+
+        /// <summary>
+        /// Creates new instance of the <see cref="ValidateOptionsResultBuilder"/> class.
+        /// </summary>
+        public ValidateOptionsResultBuilder() { }
+
+        /// <summary>
+        /// Adds validation error.
+        /// </summary>
+        /// <param name="error">Content of error message.</param>
+        /// <param name="propertyName">THe property in the option object which contains an error.</param>
+        public void AddError(string error, string? propertyName = null)
+        {
+            ThrowHelper.ThrowIfNull(error);
+            Errors.Add(propertyName is null ? error : $"Property {propertyName}: {error}");
+        }
+
+        /// <summary>
+        /// Adds any validation error carried by the <see cref="ValidationResult"/> instance to this instance.
+        /// </summary>
+        /// <param name="result">The instance to consume the errors from.</param>
+        public void AddResult(ValidationResult? result)
+        {
+            if (result?.ErrorMessage is not null)
+            {
+                Errors.Add(result.MemberNames.Any() ?
```

Comment:
Not sure what the underlying implementation of `ValidationResult.MemberNames` is in general, but do we care about enumerating it twice in this case?

Reply:
I am not expecting `ValidationResult.MemberNames` will be filled by members most of the time. Also, `Any` implementation is optimized to check if it is something like `ICollection` and return the count without enumeration. So, using `Any` would be ok.

Reply:
Is it possible that it could be an arbitrary IEnumerable like an iterator method that does side effects? If so then `Any()` could trigger any side-effects before the first `yield` keyword. But if arbitrary IEnumerables are not allowed I think it's ok.

Reply:
It is possible to have arbitrary IEnumerable, it is passed by the caller when creating the ValidationResult https://github.com/dotnet/runtime/blob/0a15c3b97dabaf94e5d4399aef2fd48c2b3f446a/src/libraries/System.ComponentModel.Annotations/src/System/ComponentModel/DataAnnotations/ValidationResult.cs#L57. My point is it will be very unlikely that users call this much and if they do, it is unlikely to use IEnumerable which is not ICollection. Also, I am not expecting to MemberNames containing more than two items.

I can rewrite the code to avoid the worst case, but I am not sure it is worth it to complicate the code to optimize for this case.

Reply:
For arbitrary enumerables we should try to avoid double enumerations. You might be able to avoid making code too complicated as follows:
```C#
string joinedMembers = string.Join(MemberSeparatorString, result.MemberNames); // for non-empty separators returns the empty string if and only if the enumerable is empty.
Errors.Add(joinedMembers.Length == 0 ? result.ErrorMessage : $"{joinedMembers}: {result.ErrorMessage}"
```